### DISCUSSION TITLE
Labelling a Partially Labelled Dataset in MONAILabel + CVAT Integration?

### DISCUSSION THREAD

**Discussion by bilalcodehub** [Project-MONAI/MONAILabel#1762] [Q&A] [answered]
Hi All,

Apologies if this is a basic question, but I want to export a partially labeled dataset to MONAILabel. I intend to use Endoscopy application. Some of the images already have labels, while others do not.  I was able to integrate the MONAILabel with CVAT. 

So my questions: what’s the best approach to export and continue labelling this dataset with MONAILabel? 

Any advice or guidance would be greatly appreciated.

Many thanks and kind regards,
Bilal

**Comment by SachidanandAlle** [comment]
Save one label in monailabel.  That will create labels/final folder in your studies.  And you can refer to that one sample label and copy other pre-existing labels over there.

**Reply by bilalcodehub** [reply]
> Save one label in monailabel. That will create labels/final folder in your studies. And you can refer to that one sample label and copy other pre-existing labels over there.

I labeled one image and saved it within the CVAT, but nothing happened on the MONAILabel server side. It seems MONAILabel only downloads labels once a task is marked as completed and finished in the CVAT menu. Is there a configuration to customise this?

**Comment by bilalcodehub** [comment]
Thanks for your response. Two followup questions:

- What happens if the task doesn’t include the original images for some of the labels that were copied to the labels/finals folder? Hope that shall be okay.
- Also, I noticed there’s an original folder inside the labels directory. Could you clarify its purpose?

**Reply by tangy5** [reply]
1. MONAILabel will create the annotation label and save to labels/finals after annotation completed, submitted. 
2.With original label, users can load existing annotation and continue to annotate or edits.

**Reply by bilalcodehub** [reply]
> 1. MONAILabel will create the annotation label and save to labels/finals after annotation completed, submitted.
>    2.With original label, users can load existing annotation and continue to annotate or edits.

Thanks. So I need to copy existing annotated masks in the labels/original folder for MONAILabel/CVAT to pick it up. Is there a specific requirement to name pre-annotated mask files?

**Reply by bilalcodehub** [reply]
> 1. 2.With original label, users can load existing annotation and continue to annotate or edits.

I copied the existing labels into the `original` folder. The MONAILabel server started without any issues and successfully created the `Project/Task` in CVAT, as specified in the configurations. However, the labels are not being overlaid on the images in CVAT, so I’m unable to refine or edit them. Is there something I need to do in CVAT to ensure the existing labels are applied to the images?

**Reply by SachidanandAlle** [reply]
Cvat and monailabel are two different services.  And such sync is not supported via monailabel.

Monailabel can only pull from cvat and it won't have a mechanism yet to push anything to cvat

**Reply by bilalcodehub** [reply]
> Cvat and monailabel are two different services. And such sync is not supported via monailabel.

Right, thanks.

> Monailabel can only pull from cvat and it won't have a mechanism yet to push anything to cvat

I thought MONAILabel can push masks like it pushes the images to the CVAT `Project/Tasks`. 

Any workaround you suggest to utilise existing labels?

**Reply by bilalcodehub** [reply]
---

**Comment by SachidanandAlle** [comment]
Currently I am trying to add SAM2 support.  While doing this, let me see if
I can address some of these situations as well.

**Reply by bilalcodehub** [reply]
> Currently I am trying to add SAM2 support. While doing this, let me see if I can address some of these situations as well.

SAM2 integration sounds excellent! We’re currently fine-tuning SAM2 on surgical videos to detect a wide range of objects. It would be great if users could upload their own weights instead of relying solely on the default SAM2 models. Any tentative release date for SAM2 integration?

Wish I could get MONAILabel + CVAT working for our custom use case discussed in the current thread. 😕

**Reply by SachidanandAlle** [reply]
Push the labels (sync) need some changes/implementation at monailabel side.  

Let me see if I can add those fixes along with SAM2.  Yes the idea is you bring in your fine tuned SAM2 for now.  Not supporting the fine-tuning of SAM2 within monailabel.

So, hopefully I will make a release candidate in a week or two.

**Reply by SachidanandAlle** [reply]
In summary you have annotations in the local folder for some of the images and you want to see them in cvat.

If I remember correctly, for your masks you can always generate the corresponding cvat annotations and using the API you can push it to cvat against a image/task.

I feel this is possibly a solved thing in the cvat community.

**Reply by bilalcodehub** [reply, marked as answer]
I finally got everything working! I wrote a script to convert masks into COCO format and uploaded them to the respective task in CVAT—it worked flawlessly. Adding an option in MONAILabel to sync labels from the original folder when creating a task would be a great time-saver for users.

I’ve successfully submitted the existing labels to the server via CVAT, and the active learning process has started training the model. For the first time, there were no errors—it’s a great day!

Thank you so much for your support and guidance throughout this process. I truly appreciate it!
